File: Editorial/94_Binary_Tree_Inorder_Traversal/TypeScript/3_Morris_Traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function inorderTraversal(root: TreeNode | null): number[] {
    let res: number[] = [];
    let curr = root;
    let pre;

    while (curr != null) {
        if (curr.left == null) {
            res.push(curr.val);
            curr = curr.right; // move to next right node
        } else {
            pre = curr.left;
            while (pre.right != null && pre.right !== curr) { // find rightmost
                pre = pre.right;
            }

            if (pre.right === null) {
                // establish a link back to the current node
                pre.right = curr;
                curr = curr.left;
            } else {
                // restore the tree structure
                pre.right = null;
                res.push(curr.val);
                curr = curr.right;
            }
        }
    }

    return res;
}

File: Editorial/94_Binary_Tree_Inorder_Traversal/Go/3_Morris_Traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func inorderTraversal(root *TreeNode) []int {
    res := []int{}
    curr := root
    var pre *TreeNode
    for curr != nil {
        if curr.Left == nil {
            res = append(res, curr.Val)
            curr = curr.Right // move to next right node
        } else { // has a left subtree
            pre = curr.Left
            for pre.Right != nil && pre.Right != curr { // find rightmost
                pre = pre.Right
            }
            if pre.Right == nil { // if the rightmost node's right child is nil
                pre.Right = curr // put current after the pre node
                curr = curr.Left // move curr to the top of the new tree
            } else { // the rightmost node's right child is not null
                pre.Right = nil             // restore the tree
                res = append(res, curr.Val) // add current value to the result list
                curr = curr.Right           // move current node to the right
            }
        }
    }
    return res
}
